### PR TITLE
feat(skipcpio): speed up and harden skipcpio

### PR DESCRIPTION
**Pull request by haraldh**
Before:
```
Benchmark #1: ./skipcpio/skipcpio test-5.10.15-200.fc33.x86_64.img >/dev/null
  Time (mean ± σ):     125.5 ms ±   0.9 ms    [User: 97.4 ms, System: 27.9 ms]
  Range (min … max):   124.8 ms … 129.4 ms    23 runs
```

After:
```
Benchmark #1: ./skipcpio/skipcpio test-5.10.15-200.fc33.x86_64.img >/dev/null
  Time (mean ± σ):      12.2 ms ±   0.3 ms    [User: 2.7 ms, System: 9.5 ms]
  Range (min … max):    11.7 ms …  13.6 ms    212 runs
```

Besides the speedup, skipcpio now parses the cpio header and is not
falsely ending when the early cpio payload contains `TRAILER!!!`.

Fixes: https://github.com/dracutdevs/dracut/issues/1123

### CODE REVIEW

File: skipcpio/skipcpio.c [hunk 1]
```diff
@@ -23,16 +23,65 @@
...
+        char c_namesize[8];
+        char c_check[8];
+};
+
+struct buf_struct {
+        struct cpio_header h;
+        char filename[CPIO_ENDLEN];
+};
+
+static struct buf_struct buf;
```
Comment (ddiss) [inline skipcpio/skipcpio.c:57]: nit: you could turn this into a union with `char b[4096]` and reuse it for copy-buffer.

File: skipcpio/skipcpio.c [hunk 2]
```diff
@@ -23,16 +23,65 @@
 #define _GNU_SOURCE
 #endif
```
Comment (ddiss) [inline skipcpio/skipcpio.c:24]: It'd probably make sense to add the libcpio copyright too, e.g.:
```

/*
 * Some code from SEl4 libcpio, which is:
 * Copyright 2017, Data61, CSIRO (ABN 41 687 119 230)
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */
```

File: skipcpio/skipcpio.c [hunk 3]
```diff
@@ -60,59 +109,82 @@ int main(int argc, char **argv)
...
                                 if (s <= 0)
                                         break;
 
-                                for (i = 0; (i < s) && (buf[i] == 0); i++) ;
+                                for (i = 0; (i < s) && (copy_buffer[i] == 0); i++) ;
 
-                                if (buf[i] != 0) {
+                                if (copy_buffer[i] != 0) {
                                         pos += i;
+                                        if (pos & 3) {
```
Comment (ddiss) [inline skipcpio/skipcpio.c:167]: IIUC, the zeros here aren't strictly part of a cpio archive, so probably shouldn't be subject to the 4 byte alignment requirement. `Documentation/driver-api/early-userspace/buffer-format.rst` indicates that kernel accepts `0 or more occurrences`of `("\0" | cpio_archive | cpio_gzip_archive)` which would imply that "\0" can be unaligned.

**Review by ddiss — APPROVED** [review]
Looks reasonable to me. A few minor comments inline.

**Review by danimo — APPROVED** [review]

**Review by johannbg — CHANGES_REQUESTED** [review]
Add the missing libcpio copyright as suggested by @ddiss

File: skipcpio/skipcpio.c [hunk 4]
```diff
@@ -60,59 +109,82 @@ int main(int argc, char **argv)
         fseek(f, 0, SEEK_SET);
 
         /* check, if this is a cpio archive */
-        if (buf[0] == '0' && buf[1] == '7' && buf[2] == '0' && buf[3] == '7' && buf[4] == '0' && buf[5] == '1') {
+        if (buf.h.c_magic[0] == '0' && buf.h.c_magic[1] == '7' &&
+            buf.h.c_magic[2] == '0' && buf.h.c_magic[3] == '7' && buf.h.c_magic[4] == '0' && buf.h.c_magic[5] == '1') {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply skipcpio/skipcpio.c:113]: true

File: skipcpio/skipcpio.c [hunk 5]
```diff
@@ -60,59 +109,82 @@ int main(int argc, char **argv)
...
+                        filename_length = parse_hex_str(buf.h.c_namesize, sizeof(buf.h.c_namesize));
+                        pos = align_up(pos + sizeof(struct cpio_header) + filename_length, CPIO_ALIGNMENT);
 
-                        h = memmem(buf, sizeof(buf), CPIO_END, sizeof(CPIO_END));
-                        if (h) {
-                                pos = (h - buf) + pos + CPIO_ENDLEN;
+                        filesize = parse_hex_str(buf.h.c_filesize, sizeof(buf.h.c_filesize));
+                        pos = align_up(pos + filesize, CPIO_ALIGNMENT);
+
+                        if (filename_length == (CPIO_ENDLEN + 1) && strncmp(buf.filename, CPIO_END, CPIO_ENDLEN) == 0) {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply skipcpio/skipcpio.c:127]: unlikely
from the `cpio` source:
```C
  file_hdr.c_namesize = 11;
  file_hdr.c_name = CPIO_TRAILER_NAME;
```

File: skipcpio/skipcpio.c (hunk 3 above)
Comment (haraldh) [reply skipcpio/skipcpio.c:167]: removed...

File: skipcpio/skipcpio.c (hunk 2 above)
Comment (haraldh) [reply skipcpio/skipcpio.c:24]: removed the copied code

**Review by johannbg — APPROVED** [review]

**Comment by haraldh** [comment]
> Add the missing libcpio copyright as suggested by @ddiss

@johannbg removed the copied code

File: skipcpio/skipcpio.c (hunk 1 above)
Comment (haraldh) [reply skipcpio/skipcpio.c]: done

**Comment by haraldh** [comment]
@ddiss please re-review

**Review by johannbg — APPROVED** [review]

File: skipcpio/skipcpio.c [hunk 6]
```diff
@@ -60,59 +90,78 @@ int main(int argc, char **argv)
...
-                                pos = (h - buf) + pos + CPIO_ENDLEN;
+                do {
+                        // zero string, spilling into next unused field, to use strtol
+                        buf.cpio.h.c_chksum[0] = 0;
+                        filename_length = strtol(buf.cpio.h.c_namesize, NULL, 16);
+                        pos = ALIGN_UP(pos + sizeof(struct cpio_header) + filename_length, CPIO_ALIGNMENT);
+
+                        // zero string, spilling into next unused field, to use strtol
+                        buf.cpio.h.c_dev_maj[0] = 0;
+                        filesize = strtol(buf.cpio.h.c_filesize, NULL, 16);
```
Comment (ddiss) [inline skipcpio/skipcpio.c:108]: This should be `strtoul()` (same for comment above). looks fine otherwise.

Comment (haraldh) [reply skipcpio/skipcpio.c:108]: you are so right.. thanks

Comment (haraldh) [reply skipcpio/skipcpio.c]: done

**Review by johannbg — APPROVED** [review]

**Pull request merged**